blaze-client/src/test/scala/org/http4s/client/blaze/ClientTimeoutSpec.scala

```diff
@@ -156,5 +157,29 @@ class ClientTimeoutSpec extends Http4sSpec {
 
       c.fetchAs[String](FooRequest).unsafeRunSync must_== "done"
     }
+
+    // Regression test for: https://github.com/http4s/http4s/issues/2386
+    // and https://github.com/http4s/http4s/issues/2338
+    "Eventually timeout on connect timeout" in {
+      val manager = ConnectionManager.basic[IO, BlazeConnection[IO]]({ _ =>
+        // In a real use case this timeout is under OS's control (AsynchronousSocketChannel.connect)
+        IO.sleep(2.seconds) *> IO.raiseError[BlazeConnection[IO]](new IOException())
```

Comment:
It ate my comment.

What are the actual requirements for sleeping here?  Somewhat longer than `requestTimeout`, less than the argument to `unsafeRunTimed`?  I'm concerned about blocking a thread for two seconds, which slows down test runs and bogs other things down.  I think the way we do this contributes to our instability in blaze tests on Travis CI.

The trick is that if the timeout is too aggressive, we run into race conditions, and if it's too slow, we have slow CI and can starve other tests.  If we were smarter about `cats.effect.Timer`, we'd have less problems like this, but that would require more plumbing into blaze.  That's well beyond the scope of this ticket.

Anyway, I'd like to make these timeouts as tight as we can without running into noisy failures.  Just making sure I understand the ordering.

Reply:
The requirement is that `requestTimeout` < `IO.sleep` < `unsafeRunTimed`. For as long as everything works this tests will run for the time of `IO.sleep`, so this it the number that we should care the most about. `requestTimeout` can be near zero. `unsafeRunTimed` affects only run time of failing test, so we can be a bit more generous with it.

Maybe we could go with something like `requestTimeout = 50.millis`, `IO.sleep(1.second)`, `unsafeRunTimed(2.seconds)` ? That would make this test case the fastest in the `ClientTimeoutSpec`.